### ISSUE TITLE
Typed labels for metrics

### ISSUE THREAD

**Issue by vigoo** [zio/zio-zmx#206]
I prefer having well typed labels for my metrics.

A minimum implementation could be to be to have wrappers on top of the existing "untyped" metric reporter effects, something like:

```scala
def count(name: String, v: Double, tags: (String, String)*): ZIO[Any, Nothing, Unit]

class Counter[T](name: String, tagMapper: T => Seq[String, String]) {
  def inc(value: Double, tags: T): ZIO[Any, Nothing, Unit] = count(name, value, tagMapper(tags))
}
```

this already can be used by manually providing the mapping like

```scala
val c1 = new Counter[(String, UUID)]("counter1", { case (step, id) => Seq("step" -> step, "id" -> id.toString) })
c1.inc(1.0)
```

Then it could be made more convenient by 

- being able to define the mapping with a typeclass
- helper functions to create the typed metrics and to define the mappings/mapping type classes

**Comment by toxicafunk** [comment]
We have a branch pending to be merged on ZIO-Metrics that adds statically typed labels for prometheus which might be of interest here :

https://github.com/zio/zio-metrics/blob/prometheus2/prometheus/src/main/scala/zio/metrics/prometheus/Metric.scala

**Comment by adamgfraser** [comment]
Yes this is a great idea and will also support a more efficient implementation of capturing the underlying metrics in some cases since the class can directly contain a reference to the value that needs to be updated instead of having to look it up in a map. Will incorporate into #209.

**Comment by adamgfraser** [comment]
Closed by #209.

**Comment by adamgfraser** [comment]
Just getting down some thoughts on this. The most idiomatic way to work with metrics is to use metric aspects and you would typically define an object with all the relevant metrics to your application or module. For example:

```scala
object Metrics {

  val countFollowerRequests: MetricAspect[Any] =
    MetricAspect.countSuccesses("followerRequests")
}
```

As written this does not involve typed labels (in fact it doesn't expose labels at all outside of this object, but if the user desires to use typed labels they can do so like this:

```scala
final case class CompanyName(name: String)
final case class ClusterId(id: Int)
​
object Metrics {
  def countWith(companyName: CompanyName, clusterId: ClusterId): MetricAspect[Any] =
    MetricAspect.countSuccesses(???) // compute metric name and labels based on typed inputs here
}
```

This allows defining further metrics in terms of this one in a fully typed way.

This still relies on some level of discipline for the developer to define metrics in one place and to not map different typed values to the same underlying labels, but that seems unavoidable.

Things we could potentially improve on would include better documentation about the "define once, use everywhere" philosophy and possibly making the metrics themselves as opposed to the metric aspects less prominent in the API.

One other thing that could be worth thinking about is whether we should do some further translation between user provided name and labels and the name and labels we send to downstream systems. ZIO ZMX itself guarantees that metrics can never "clash" with each other, regardless of names, because the `MetricKey` is a unique key that reflects all information that goes into defining that metric. But I don't think downstream services honor the same property. One simple solution would be to prepend the metric
